pcap-dpdk.c

```diff
@@ -0,0 +1,800 @@
+
+struct dpdk_ts_helper{
+	struct timeval start_time;
+	uint64_t start_cycles;
+	uint64_t hz;
+};
+struct pcap_dpdk{
+	pcap_t * orig;
+	uint16_t portid; // portid of DPDK
+	pcap_handler cb; //callback and argument
+	u_char *cb_arg;
+	int max_cnt;
+	int must_clear_promisc;
+	int filter_in_userland;
+	uint64_t rx_pkts;
+	uint64_t bpf_drop;
+	struct ether_addr eth_addr;
+	char mac_addr[DPDK_MAC_ADDR_SIZE];
+	struct timeval prev_ts;
+	struct rte_eth_stats prev_stats;
+	struct timeval curr_ts;
+	struct rte_eth_stats curr_stats;
+	uint64_t pps;
+	uint64_t bps;
+	struct rte_mempool * pktmbuf_pool;
+	struct dpdk_ts_helper ts_helper;
+	char pci_addr[DPDK_PCI_ADDR_SIZE];
+	unsigned char pcap_tmp_buf[RTE_ETH_PCAP_SNAPLEN];
+};
+
+static struct rte_eth_conf port_conf = {
+	.rxmode = {
+		.split_hdr_size = 0,
+	},
+	.txmode = {
+		.mq_mode = ETH_MQ_TX_NONE,
+	},
+};
+
+static int dpdk_init_timer(struct pcap_dpdk *pd){
+	gettimeofday(&(pd->ts_helper.start_time),NULL);
+	pd->ts_helper.start_cycles = rte_get_timer_cycles();
+	pd->ts_helper.hz = rte_get_timer_hz();
+	if (pd->ts_helper.hz == 0){
+		return -1;	
+	}
+	return 0;
+}
+static inline void calculate_timestamp(struct dpdk_ts_helper *helper,struct timeval *ts)
+{
+	uint64_t cycles;
+	// delta
+	struct timeval cur_time;
+	cycles = rte_get_timer_cycles() - helper->start_cycles;
+	cur_time.tv_sec = (time_t)(cycles/helper->hz);
+	cur_time.tv_usec = (suseconds_t)((cycles%helper->hz)*1e6/helper->hz);
+	timeradd(&(helper->start_time), &cur_time, ts);
+}
+
+static uint32_t dpdk_gather_data(unsigned char *data, int len, struct rte_mbuf *mbuf)
+{
+	uint32_t total_len = 0;
+	while (mbuf && (total_len+mbuf->data_len) < len ){
+		rte_memcpy(data+total_len, rte_pktmbuf_mtod(mbuf,void *),mbuf->data_len);
+		total_len+=mbuf->data_len;
+		mbuf=mbuf->next;
+	}
+	return total_len;
+}
+
+static void dpdk_dispatch_internal(void *dpdk_user)
+{
+	if (dpdk_user == NULL){
+		return;
+	}
+	pcap_t *p = dpdk_user;
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	int max_cnt = pd->max_cnt;
+	int burst_cnt = 0;
+	pcap_handler cb = pd->cb;
+	u_char *cb_arg = pd->cb_arg;
+	int nb_rx=0;
+	struct rte_mbuf *pkts_burst[MAX_PKT_BURST];
+	struct rte_mbuf *m;
+	struct pcap_pkthdr pcap_header;
+	uint16_t portid = pd->portid;
+	unsigned lcore_id = rte_lcore_id();
+	unsigned master_lcore_id = rte_get_master_lcore();
+	// In DPDK, pkt_len is sum of lengths for all segments. And data_len is for one segment
+	uint16_t data_len = 0;
+	uint32_t pkt_len = 0;
+	int caplen = 0;
+	u_char *bp = NULL;
+	int i=0;
+	unsigned int gather_len =0;
+	int pkt_cnt = 0;
+	int is_accepted=0;
+	u_char *large_buffer=NULL;
+		
+	if(lcore_id == master_lcore_id){
+		RTE_LOG(DEBUG, USER1, "dpdk: lcoreid=%u runs for portid=%u\n", lcore_id, portid);
+	}else{
+		RTE_LOG(DEBUG, USER1, "dpdk: lcore %u has nothing to do\n", lcore_id);
+	}
+	//only use master lcore
+	if (lcore_id != master_lcore_id){
+		return;
+	}
+	if (max_cnt>0 && max_cnt < MAX_PKT_BURST){
```

Comment:
You could also use `PACKET_COUNT_IS_UNLIMITED()` here:

    if (! PACKET_COUNT_IS_UNLIMITED(max_cnt) && max_cnt < MAX_PKT_BURST){

Reply:
OK. Thanks.